Introduce nginx container and run in a WSGI server

giorgiosironi opened the pull request:
Continuation of https://github.com/libero/libero/issues/96 towards a production-like environment.

Commit:
Rename service to app

Commit:
Introduce dev image with git installed

Commit:
Add more build dependencies to dev container

Commit:
Install uwsgi to use it as a server

Commit:
Run uwsgi in prod container

Commit:
Add nginx container

Commit:
Simplicy uwsgi-related configurations

Commit:
Log to stderr

Commit:
Document uwsgi configuration with an eye to production usage

Commit:
Match nginx and uwsgi timeouts

Commit:
Remove again targets in favor of build argument

giorgiosironi commented on docker/Dockerfile:
```diff
@@ -25,12 +25,14 @@ RUN apk add --no-cache --virtual .build-dependencies \
     && \
     poetry install ${dev} && \
-    apk del .build-dependencies
+    apk del ${simulate_removal} .build-dependencies
```
Long story short: if the `dev` image removes the packages in `.build-dependencies`, `poetry install` may break as part of `make add-dependency` as it doesn't find `git` or can't compile other packages.

GiancarloFusiello commented on docker/nginx/default.conf:
```diff
@@ -0,0 +1,10 @@
+server {
+    server_name search--dev;
```
Will nginx be used for dev purposes?

giorgiosironi replied:
No, in a `dev` environment (a person with a laptop) the Flask standalone server should be enough to run code and debug it. What really is important is to have in Continuous Integration an environment that is production-like, so eventually we will have in the build some smoke tests that try to load `/ping` or other APIs through the nginx and WSGI app server combination.

GiancarloFusiello replied:
Can these tests be done on the actual production search container before deployment?

giorgiosironi replied:
Conflating image and container here, but usually we do:
- run smoke tests on the _image_ when put into a production-like environment like `ci` or `end2end` or anything else that is available for testing. This finds problem in the image itself.
- run smoke tests on the actual _container_ in the real production environment, _after_ deployment (because you need to start the container with the new image to be able to do it). This finds problem in the configuration being used.

So the answer is: yes, but we don't have those tests yet. I'll add them to this PR and cover the first of these two points.

GiancarloFusiello commented on docker/uwsgi.ini:
```diff
@@ -0,0 +1,15 @@
+# master and workers mode
+master=True
+# process pool size
+processes=2
```
Can this be inferred by the number of cores? If so can we do: `processes = number_of_cores - 1`

giorgiosironi replied:
I have no stake in the actual values here as it's mostly for demonstration on how to run the project in a production-like environment. The `sample-configuration` repository will have its copy of this which may use different values.

I'll check what uWSGI supports or uses as a default, but I'd reserve tuning for the actual production configuration: at the moment cores would be shared between different containers anyway as they are all deployed to a single server.

giorgiosironi replied:
(I wanted this to be >1 to be more realistic and discover possible concurrency problems)

giorgiosironi replied:
The default is 1 so not good for the previous comment. Only integers are allowed and `0` actually breaks it, so I'll stick with 2 (or a larger value if we want) for the purpose of Continuous Integration (which doesn't use this yet, but will for smoke tests).

GiancarloFusiello commented on poetry.lock:
```diff
@@ -206,7 +206,7 @@ python-versions = "*"
 
 [[package]]
-category = "dev"
+category = "main"
```
I don't know why but Poetry keeps changing this to be a main dependency but it's a dev dependency. I raised an issue on the Poetry repo some weeks ago but I've not heard back.

giorgiosironi replied:
I'll change this manually to `dev` again (assuming there's no checksum on the lock file preventing manual changes).

GiancarloFusiello replied:
no it should be fine to change manually. Just not ideal.

GiancarloFusiello commented on Makefile:
```diff
@@ -35,7 +35,7 @@ stop:
 
 checks:
-	$(DOCKER_COMPOSE) run --rm web /bin/bash -c "\
+	$(DOCKER_COMPOSE) run --rm app /bin/bash -c "\
```
At the top of this file there are some variables. One of them is `PROJECT_CODEBASE_DIR = "search/"`. We could make another called `PROJECT_NAME = "search"` and update the other to be `PROJECT_CODEBASE_DIR = "$(PROJECT_NAME)/"` or just create `SERVICE_NAME = "app"` and use this variable throughout the file?

giorgiosironi replied:
Yes, `SERVICE_NAME` changing is an indication it could be a variable.

Commit:
Correct libero-schemas as a dev dependency

Commit:
Extract SERVICE_NAME

GiancarloFusiello commented on Makefile:
```diff
@@ -50,38 +51,38 @@ checks:
 .PHONY: run
 run:
-	$(DOCKER_COMPOSE) run --rm --service-ports web
+	$(DOCKER_COMPOSE) run --rm --service-ports app
```
needs `app` to be replaced with `$(SERVICE_NAME)`

giorgiosironi replied:
(weird find and replace)

GiancarloFusiello commented on Makefile:
```diff
@@ -50,38 +51,38 @@ checks:
 
 build:
-	$(DOCKER_COMPOSE) build web
+	$(DOCKER_COMPOSE) build app
```
needs `app` to be replaced with `$(SERVICE_NAME)`

GiancarloFusiello commented on docker/docker-compose.yml:
```diff
@@ -1,10 +1,22 @@
 
 services:
-  web:
+  app:
```
do you think we can change the name from `app` to `search` or `libero-search` and `web` to `nginx`?

giorgiosironi replied:
Looking at https://github.com/libero/browser/blob/master/docker-compose.yaml we tend to go for role more than for implementation, for example you could replace Nginx with Apache (hypothetical) without the service name changing.
In some cases I renamed `app` to `wsgi` which is the actual role (listening for connections using the WSGI-protocol), but the service `app` remains useful for `docker-compose run app python search/something.py` so it's not true that it's only WSGI-related if it can be used for command line processes.

https://github.com/libero/search/pull/8/commits/dcad5e75c3099282f137be239bb6efe917a55fb5 now forces explanatory container names such as `search_app_1`.

GiancarloFusiello replied:
I like this change. Does docker know to look for `.env` automagically?

giorgiosironi replied:
[`docker-compose` does](https://docs.docker.com/compose/env-file/): the values are either used directly by it or available for interpolation in `docker-compose.yml` files.

GiancarloFusiello replied:
I ask because there is a .env file in the `docker/` directory. Can we repurpose this?

GiancarloFusiello replied:
Ideally, I don't want files in the root directory if it can be avoided.

giorgiosironi replied:
They are not related apart from the extension; `docker-compose` won't read `dev.env` automatically (because of the name) and it is [explicitly referenced](https://github.com/libero/search/blob/master/docker/docker-compose.dev.yml#L17) for its variables to be passed to the `app` service containers. You don't want `COMPOSE_PROJECT_NAME` to be passed to the Python container (it doesn't need it) and at the same time you don't need `FLASK_ENV` to be available for interpolation in `docker-compose.yml`.

GiancarloFusiello replied:
So the `.env` file needs to be in the root directory or can it live in the same directory as the docker-compose file?

giorgiosironi replied:
> Ideally, I don't want files in the root directory if it can be avoided.

Unfortunately the documentation (just linked) doesn't provide for a way to customize the path to it, and we would have to execute `docker-compose` exclusively from inside the `docker/` directory which would be annoying (and error-prone as it won't produce an error if run from the root, just ignore a few configurations).

GiancarloFusiello replied:
yeah that's not ideal. Leave it where it is if that's the intended design.

giorgiosironi replied:
(since it's needed for rendering the `docker-compose.yml`, it's probably read before that and hence there's no place to set a configuration for its path)

Commit:
Add smoke test to build

Commit:
Replace missing SERVICE_NAME references

Commit:
Force consistent container names

```
$ docker-compose -f docker/docker-compose.yml ps
Name              Command          State          Ports
-------------------------------------------------------------------
search_app_1   uwsgi --ini=uwsgi.ini   Up
search_web_1   nginx -g daemon off;    Up      0.0.0.0:8080->80/tcp
```

giorgiosironi commented:
Smoke test selected output:
```
$ .travis/smoke-test.sh
Pulling web (nginx:1.15.5-alpine)...
1.15.5-alpine: Pulling from library/nginx
Digest: sha256:ae5da813f8ad7fa785d7668f0b018ecc8c3a87331527a61d83b3b5e816a0f03c
Status: Downloaded newer image for nginx:1.15.5-alpine
Creating docker_app_1 ... 
Creating docker_app_1
Creating docker_web_1
app_1  | [uWSGI] getting INI configuration from uwsgi.ini
app_1  | *** Starting uWSGI 2.0.18 (64bit) on [Fri Mar  1 10:40:41 2019] ***
web_1  | 172.18.0.1 - - [01/Mar/2019:10:40:42 +0000] "GET /ping HTTP/1.1" 200 4 "-" "curl/7.35.0" "-"
app_1  | [pid: 5|app: 0|req: 1/1] 172.18.0.1 () {32 vars in 353 bytes} [Fri Mar  1 10:40:42 2019] GET /ping => generated 4 bytes in 2 msecs (HTTP/1.1 200) 3 headers in 121 bytes (1 switches on core 0)
```

GiancarloFusiello commented:
@giorgiosironi Do you plan on adding anything else to this PR?

giorgiosironi commented:
The smoke tests are there and running, so there a couple of TODOs that I'll deal with later (don't want to add even more complexity in one shot).
